File: CPP_FFFE/21_ReferenceAndValueSemantics/code/04_passing.cpp
#include <iostream>

struct Fraction {
  int numerator, denominator;
  auto operator<=>(Fraction const &) const = default;
  friend std::ostream & operator<<(std::ostream &, Fraction const &);
};
std::ostream & operator<<(std::ostream & os, Fraction const & f) {
  os << "(" << f.numerator << "/" << f.denominator << ")";
  return os;
}

void change_fraction(Fraction f) {
  f.numerator = 1;
  f.denominator = 2;
  std::cout << "Inside change_fraction, f == " << f << "\n";
}

void make_new_fraction(Fraction& f) {
  f = {8, 13}; // calls Fraction constructor
  std::cout << "Inside make_new_fraction, f == " << f << "\n";
}

int main() {
  Fraction a {3, 5};
  std::cout << "Before change_fraction, a == " << a << "\n";
  change_fraction(a);
  std::cout << "After change_fraction, a == " << a << "\n\n";

  std::cout << "Before make_new_fraction, a == " << a << "\n";
  make_new_fraction(a);
  std::cout << "After make_new_fraction, a == " << a << "\n";
}

File: CPP_FFFE/21_ReferenceAndValueSemantics/code/04_passing.java
class Fraction {
  public int numerator, denominator;
  public Fraction(int n, int d) {
    numerator = n;
    denominator = d;
  }
}

class Passing {
  static void change_fraction(Fraction f) {
    f.numerator = 1;
    f.denominator = 2;
    System.out.printf("Inside change_fraction, f == (%d/%d)\n", f.numerator, f.denominator);
  }

  static void make_new_fraction(Fraction f) {
    f = new Fraction(8, 13);
    System.out.printf("Inside make_new_fraction, f == (%d/%d)\n", f.numerator, f.denominator);
  }

  public static void main(String[] args) {
    Fraction a = new Fraction(3, 5);
    System.out.printf("Before change_fraction, a == (%d/%d)\n", a.numerator, a.denominator);
    change_fraction(a);
    System.out.printf("After change_fraction, a == (%d/%d)\n", a.numerator, a.denominator);
    System.out.println();
    System.out.printf("Before make_new_fraction, a == (%d/%d)\n", a.numerator, a.denominator);
    make_new_fraction(a);
    System.out.printf("After make_new_fraction, a == (%d/%d)\n", a.numerator, a.denominator);
  }
}
